CONTRIBUTING.md

```diff
@@ -1,69 +1,138 @@
+which will run `cargo make ci` before each push. To initialize the hook  run `cargo test`. If
+`cargo-make` is not installed, it will provide instructions to install it for you. This will ensure
+that your code is formatted, compiles and passes all tests before you push. If you need to skip this
+check, you can use `git push --no-verify`.
+
+### Sign your commits
+
+We use commit signature verification, which blocks your commits from being merged via the UI unless
+you have setup your machine to sign commits. See
+<https://docs.github.com/en/authentication/managing-commit-signature-verification/about-commit-signature-verification>
+for more details.
 
 ## Implementation Guidelines
 
-### Use of unsafe for optimization purposes
+### Setup
 
-**Do not** use unsafe to achieve better performances. This is subject to change, [see.](https://github.com/tui-rs-revival/tui-rs-revival/discussions/66)
-The only exception to this rule is if it's to fix **reproducible slowness.**
+Clone the repo and build it using [cargo-make](https://sagiegurari.github.io/cargo-make/)
 
-## Building
+`ratatui` is an ordinary Rust project where common tasks are managed with
+[cargo-make](https://github.com/sagiegurari/cargo-make/). It wraps common `cargo` commands with sane
+defaults depending on your platform of choice. Building the project should be as easy as running
+`cargo make build`.
 
-[cargo-make]: https://github.com/sagiegurari/cargo-make "cargo-make"
+```shell
+git clone https://github.com/tui-rs-revival/ratatui.git
+cd ratatui
+cargo make build
+```
 
-`ratatui` is an ordinary Rust project where common tasks are managed with [cargo-make].
-It wraps common `cargo` commands with sane defaults depending on your platform of choice.
-Building the project should be as easy as running `cargo make build`.
+### Tests
 
-## :hammer_and_wrench: Pull requests
+The [test coverage](https://app.codecov.io/gh/tui-rs-revival/ratatui) of the crate is reasonably
+good, but this can always be improved. Focus on keeping the tests simple and obvious and write unit
+tests for all new or modified code. Beside the usual doc and unit tests, one of the most valuable
+test you can write for `ratatui` is a test against the `TestBackend`. It allows you to assert the
+content of the output buffer that would have been flushed to the terminal after a given draw call.
+See `widgets_block_renders` in [tests/widgets_block.rs](./tests/widget_block.rs) for an example.
 
-All contributions are obviously welcome.
-Please include as many details as possible in your PR description to help the reviewer (follow the provided template).
-Make sure to highlight changes which may need additional attention or you are uncertain about.
-Any idea with a large scale impact on the crate or its users should ideally be discussed in a "Feature Request" issue beforehand.
+Prefer unit tests (and doc tests) in the code file over integration tests.
```

Comment:
Not a syntactically valid sentence :P

Reply:
It actually is syntactically correct (imperative mood form of "prefer" with an implied but missing -"to create"), but I'll reword this to something like "When writing tests, generally prefer to unit and doc tests within the file that is being tested rather than integration tests in the `tests/` folder"